### PR TITLE
BTFS-1166

### PR DESCRIPTION
**Pull request by laipogo**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
refactoring

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior?** (You can also refer to a JIRA ticket here)
switch get hosts from http to grpc

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **What dependencies / modules need to be updated as pre-requisites?** (Include relevant PR links here)

* **Description of changes**

---

* **Please check if the PR fulfills these requirements**
- [x] The subject of this PR contains a JIRA ticket BTFS-xxxx (N/A for community PRs)
- [ ] PR description and commit messages are [good and meaningful](https://chris.beams.io/posts/git-commit/)
- [ ] Unit tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **Please make sure the following procedures have been applied before requesting reviewers**
- [x] Code changes closely follow [Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments)
- [x] Code has been *rebased* against latest master (`git merge` not recommended, unless you know what you are doing)
- [x] Code changes have run through `go fmt`
- [x] Code changes have run through `go mod tidy`
- [x] All unit tests passed locally (`make test_go_test`)

### CODE REVIEW

**Comment by laipogo** [comment]
TestDhtFindPeer test failed.

File: core/hub/sync.go [hunk 1]
```diff
@@ -35,31 +30,32 @@ func CheckValidMode(mode string) error {
...
 	if err != nil {
-		return nil, fmt.Errorf("Failed to query BTFS-Hub service: %v", err)
-	}
-
-	var hq hostsQuery
-	if err := json.NewDecoder(resp.Body).Decode(&hq); err != nil {
 		return nil, err
 	}
+	var resp *hubpb.HostsResp
+	err = grpc.HubQueryClient(config.Services.HubDomain).WithContext(context.Background(), func(ctx context.Context,
```
Comment (taiyangc) [inline core/hub/sync.go]: We have ctx now, so can you make `QueryHub` take ctx from caller and pass it down

File: core/commands/storage/hosts_test.go [hunk 2]
```diff
@@ -31,7 +33,9 @@ func TestHostsSaveGet(t *testing.T) {
 			t.Fatal(err)
 		}
 		for i, sn := range stored {
-			if *sn != *nodes[i] {
+			bs1, _ := proto.Marshal(sn)
+			bs2, _ := proto.Marshal(nodes[i])
```
Comment (taiyangc) [inline core/commands/storage/hosts_test.go]: Check err, you can `t.Fatal` on err

File: core/commands/storage/hosts_test.go [hunk 3]
```diff
@@ -31,7 +33,9 @@ func TestHostsSaveGet(t *testing.T) {
...
+			if hex.EncodeToString(bs1) != hex.EncodeToString(bs2) {
```
Comment (taiyangc) [inline core/commands/storage/hosts_test.go]: can do `bytes.Compare` here.

File: core/hub/sync.go (hunk 1 above)
Comment (laipogo) [reply core/hub/sync.go]: sure, that's the right way.

File: core/commands/storage/hosts_test.go (hunk 2 above)
Comment (laipogo) [reply core/commands/storage/hosts_test.go]: cool

File: core/commands/storage/hosts_test.go [hunk 4]
```diff
@@ -33,9 +33,15 @@ func TestHostsSaveGet(t *testing.T) {
...
-			if hex.EncodeToString(bs1) != hex.EncodeToString(bs2) {
+			bs1, err := proto.Marshal(sn)
+			if err != nil {
+				t.Fatal(err)
+			}
+			bs2, err := proto.Marshal(nodes[i])
+			if err != nil {
+				t.Fatal(err)
+			}
+			if bytes.Compare(bs1, bs2) == 0 {
```
Comment (taiyangc) [inline core/commands/storage/hosts_test.go]: != 0

**Pull request merged**